lightyear/src/connection/steam/client.rs

```diff
@@ -0,0 +1,159 @@
+use steamworks::networking_sockets::{NetConnection, NetworkingSockets};
+use steamworks::networking_types::{
+    NetConnectionEnd, NetConnectionInfo, NetworkingConnectionState, SendFlags,
+};
+use steamworks::ClientManager;
+use tracing::info;
+
+use super::SingleClientThreadSafe;
+
+const MAX_MESSAGE_BATCH_SIZE: usize = 512;
+
+#[derive(Debug, Clone)]
+pub struct SteamConfig {
+    pub server_addr: SocketAddr,
+    pub app_id: u32,
+}
+impl Default for SteamConfig {
+    fn default() -> Self {
+        Self {
+            server_addr: SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), 27015)),
+            // app id of the public Space Wars demo app
+            app_id: 480,
+        }
+    }
+}
+
+pub struct Client {
+    client: steamworks::Client<ClientManager>,
+    single_client: SingleClientThreadSafe,
+    config: SteamConfig,
+    connection: Option<NetConnection<ClientManager>>,
+    packet_queue: VecDeque<ReadWordBuffer>,
+}
+
+impl Client {
+    pub fn new(config: SteamConfig) -> Result<Self> {
+        let (client, single) = steamworks::Client::init_app(config.app_id)
+            .context("could not initialize steam client")?;
+        Ok(Self {
+            client,
+            single_client: SingleClientThreadSafe(Arc::new(RwLock::new(single))),
+            config,
+            connection: None,
+            packet_queue: VecDeque::new(),
+        })
+    }
+
+    fn connection_info(&self) -> Option<Result<NetConnectionInfo>> {
+        self.connection.as_ref().map(|connection| {
+            self.client
+                .networking_sockets()
+                .get_connection_info(connection)
+                .map_err(|err| anyhow!("could not get connection info"))
+        })
+    }
+
+    fn connection_state(&self) -> Result<NetworkingConnectionState> {
+        self.connection_info()
+            .unwrap_or(Err(anyhow!("no connection")))
+            .map_or(Ok(NetworkingConnectionState::None), |info| info.state())
+            .context("could not get connection state")
+    }
+}
+
+impl NetClient for Client {
+    fn connect(&mut self) -> Result<()> {
+        let options = vec![];
+        self.connection = Some(
+            self.client
+                .networking_sockets()
+                .connect_by_ip_address(self.config.server_addr, options)
+                .context("failed to create connection")?,
+        );
+        info!(
+            "Opened steam connection to server at address: {}",
+            self.config.server_addr
+        );
+        Ok(())
+    }
+
+    fn is_connected(&self) -> bool {
+        matches!(
+            self.connection_state(),
+            Ok(NetworkingConnectionState::Connected)
+        )
+    }
+
+    fn try_update(&mut self, delta_ms: f64) -> Result<()> {
+        if let Ok(single) = self.single_client.0.read() {
+            // TODO: the rest of the function probably shouldn't run when we can't get the lock?
+            single.run_callbacks();
```

Comment:
This should probably not be automatically called from Lightyear. If the user uses other Steam features then they might end up calling this multiple times per frame

Reply:
The docs say that this should be fine to be called once per render frame: https://partner.steamgames.com/doc/api/steam_api#SteamAPI_RunCallbacks

Or is the issue that `run_callbacks` does both:
- forward connection events for the networking_sockets api
- run user-defined callbacks (which the user might want to have control over when to run)?

Reply:
I see the bevy_steamworks also runs automatically once per frame, so I might just keep it that way for now until I understand steam callbacks better: https://github.com/HouraiTeahouse/bevy_steamworks/blob/main/src/lib.rs#L246

Reply:
I mean that the user might already be running callbacks as part of e.g bevy_steamworks or by using some other features from Steam, so this would effectively run callbacks multiple times.

Reply:
I don't see why someone would be using `bevy_steamworks` when he uses `lightyear`?
But `lightyear` probably should expose the steam client so that the user can also access the steam api whenever he wants, without the need of creating his own client. but in a way that he has no access to the `SingleClient` maybe? So he can't run `.run_callbacks`?

Reply:
I think for now the simplest is to just merge what I have now.
What I expose to clients currently is mostly `dyn NetClient` (https://github.com/cBournhonesque/lightyear/blob/8f3ba758080e5952f26c4b91f4d6a8a1a2422787/lightyear/src/connection/client.rs#L121) so there's no steam-specific information, but I can think about how to improve things in a future PR.

I'll also want to investigate the steam P2P apis in a different PR.

Reply:
> I don't see why someone would be using bevy_steamworks when he uses lightyear

@zwazel Steam offers a variety of features beyond networking, and as such it makes a lot of sense for the Steam client and its update schedule to be owned outside of any networking library.